cookie-authentication.md

```diff
@@ -0,0 +1,145 @@
+
+3. The integration collects the access token and refresh token from the SDK
+    response in the previous step.  It then responds to the browser request by
+    settings these values in cookies.
+
+4. Subsequent requests by the browser will supply these cookies which contain the
+    access and refresh tokens.  The integration should perform the following
+    logic, using the SDK's relevant JWT and OAuth authenticators:
+
+  * If the access token is valid, authenticate the request
+  * If the access token is invalid, attempt to use the refresh token to get a
+  new access token
+  * If a new access token is obtained, authenticate the request and store this
+  new token value in the access token cookie
+  * If a new access token cannot be obtained, reject the request and delete the
+    refresh token cookie
+
+### Rejecting Requests
+
+If the request cannot be authenticated, how to respond will depend on the
+accept type of the request:
+
+`Accept: text/html` - respond with a 302 redirect to the login view, with
+the requested url provided as the `?next=` parameter
+
+`Accept: application/json` - respond with `401 Unauthorized` and an empty
+body
+
+### Cookie Flags
+
+For each cookie, access token and refresh token, use these rules to determine
+the flags to set on the cookies when you are sending them to the browser:
+
+* The `expires` flag should be determined by the `ttl` value for the
+token, as defined by the Oauth Policy of the Stormpath Application.
+
+* The `HttpOnly` flag should be set as follows:
+
+  * If `web.<accessTokenCookie|refreshTokenCookie>.httpOnly` is false, do not set this flag
+  * Othwerwise, set this flag to true
+
+* The `path` flag should be set as follows:
+
+  * If `web.<accessTokenCookie|refreshTokenCookie>.path` is defined, use that value
+  * Othwerwise, use `web.basePath` if defined
+  * Finally, fallback to "/" as the value
+
+* The `domain` flag should be set as follows:
+
+  * If `web.<accessTokenCookie|refreshTokenCookie>.domain` is defined, use that value
+  * Othwerwise, use `web.domain` if defined
+  * Finally, fallback to "" as the value
+
+* The `secure` flag should be set as follows:
+
+  * If `web.<accessTokenCookie|refreshTokenCookie>.secure` is defined, use that value
+  * Othwerwise, the framework should make a best effort to determine if the
+    incoming request is HTTPS and set the `secure` flag if so.
+
+## <a name="Configuration Options"></a> Configuration Options
+
+The developer has control over the name of the cookie and the flags that
+control the cookie behavior in the browser.  The lifetime of the cookies is
+controlled by the application's OAuth Policy, and as such is not represented
+in these option blocks.  But you will need to know what those policy settings
+are, and you will find them at `web.oauth2.password`.
+
+```yaml
+web:
+  accessTokenCookie:
+    name: "access_token"
+    httpOnly: true
+    secure: null
+    path: "/"
```

Comment:
`path: null` (per comments above about `contextPath`).  Our heuristics in a node app can fall back to `/` if nothing else could be found - no need to set it in the config file.